### PR TITLE
Promisify action creator return type for WP data dispatch

### PR DESCRIPTION
**Pull request by noahtallen**
## What?
I've been working on integrating a `@wordpress/data` update into a different repo. There is an existing code snippet like this:

```ts
const { doSomething } = useDispatch( storeDescriptor );
doSomething().then( doSomethingElse );
```

However, this fails because `.then` does not exist on `doSomething` (e.g. the action creator doesn't return a promise)

## Why?
According to the documentation, this is a valid use case. However, our types prevent this because `doSomething` retains its original return type and isn't wrapped with a promise.

This was also fixed in the DT types, but that fix never made it here (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60693)

## How?
I'm attempting to add a wrapper to our `ActionCreatorsOf` type, which is used to generate return types for `dispatch`. This wrapper basically says that every action creator now returns a `Promise<void>`. 

Two things this doesn't handle:
- Generator action creators. (Should we handle this?)
- Is something other than `Promise<void>` ever returned, or is it always `void`?

## Testing Instructions
unsure

### Testing Instructions for Keyboard
n/a

## Screenshots or screencast 
n/a

### CODE REVIEW

File: packages/data/src/types.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import type { combineReducers as reduxCombineReducers } from 'redux';
 
 type MapOf< T > = { [ name: string ]: T };
 
-export type ActionCreator = Function | Generator;
+export type ActionCreator = ( ...args: any[] ) => any | Generator;
```
Comment (noahtallen) [inline packages/data/src/types.ts:9]: `Function` doesn't let you add return types and isn't compatible with the arrow syntax, so it needs to be changed so that we can wrap around the type below.

File: packages/data/src/types.ts [hunk 2]
```diff
@@ -170,9 +170,17 @@ export type ConfigOf< S > = S extends StoreDescriptor< infer C > ? C : never;
...
-		? ActionCreators
+		? PromisifiedActionCreator< ActionCreators >
 		: never;
 
+// When dispatching an action creator, the return value is a promise.
+type PromisifiedActionCreator< ActionCreators extends MapOf< ActionCreator > > =
+	{
+		[ Action in keyof ActionCreators ]: (
+			...args: Parameters< ActionCreators[ Action ] >
+		) => Promise< void >;
```
Comment (jsnajdr) [inline packages/data/src/types.ts]: Instead of `Promise< void >`, should this be `Promise< ReturnType< ActionCreators[ Action ] > >`? It's more precise type, thunk actions commonly return some value.

File: packages/data/src/types.ts [hunk 3]
```diff
@@ -170,9 +170,17 @@ export type ConfigOf< S > = S extends StoreDescriptor< infer C > ? C : never;
 
 export type ActionCreatorsOf< Config extends AnyConfig > =
 	Config extends ReduxStoreConfig< any, infer ActionCreators, any >
-		? ActionCreators
+		? PromisifiedActionCreator< ActionCreators >
 		: never;
 
+// When dispatching an action creator, the return value is a promise.
+type PromisifiedActionCreator< ActionCreators extends MapOf< ActionCreator > > =
```
Comment (jsnajdr) [inline packages/data/src/types.ts]: The name should be `PromisifiedActionCreators` (plural) because it works with the entire map of actions, not just with an individual one.

`PromisifiedActionCreator` (singular) could be a nice helper that allows us to do the promisification in two parts:
```ts
type PromisifiedActionCreators< ActionCreators extends MapOf< ActionCreator > > =
  { [ Action in keyof ActionCreators ]: PromisifyActionCreator< ActionCreators[ Action ] > };

Comment (tyxla) [inline packages/data/src/types.ts]: Might be worth exporting the new types in order to be able to consume them externally. Including the singular `PromisifiedActionCreator` and the plural `PromisifiedActionCreators` that Jarda suggested.

**Review by dmsnell — APPROVED** [review]
If this matches the actual code let's do it. Thanks @noahtallen.

A couple notes though:
 - @jsnajdr and @tyxla are right: use of the singular `PromisifiedActionCreator` will help not only with code reuse, but TypeScript will give up in some anonymous cases like this where having a separated named type-level function lets it continue. I don't understand the heuristic it uses, but I've confirmed it in several different cases.
 - Let's ensure that we don't inadvertently trigger a host of "didn't wait for `Promise` to resolve" errors where code dispatches and moves on.

the more concerning thing to me is that we await the result of dispatch, but I don't suppose that has anything to do with matching the types to the reality.

File: packages/data/src/types.ts (hunk 2 above)
Comment (noahtallen) [reply packages/data/src/types.ts]: Yeah, I was thinking about this! My concern was a use case like this:

```ts
// Action creator:
function doFoo( value: number ) {
  return { type: 'FOO', value };
}

const { doFoo } = useDispatch( store );

doFoo.then( value => {
  console.log( value ); // I assume undefined, but the type is actually { type: 'FOO', value: number } 
} );
```

File: packages/data/src/types.ts (hunk 3 above)
Comment (noahtallen) [reply packages/data/src/types.ts]: Added

Comment (noahtallen) [reply packages/data/src/types.ts]: Done!

File: packages/data/src/types.ts (hunk 2 above)
Comment (noahtallen) [reply packages/data/src/types.ts]: Thoughts on how to handle that? Can we tell if the action is a thunk? (Would those be generators?)

**Comment by noahtallen** [comment]
Here is the key line in the code

https://github.com/WordPress/gutenberg/blob/c355d5472e2d9c4c5d611f02b5b74b5efb6957a4/packages/data/src/redux-store/index.js#L200-L203

I think redux itself is synchronous here, but we wrap `Promise.resolve` around it. @jsnajdr is this just a way to make the API consistent with async thunks?

File: packages/data/src/types.ts (hunk 2 above)
Comment (noahtallen) [reply packages/data/src/types.ts]: Actually, after reading the [redux docs](https://redux.js.org/api/store#dispatchaction), the return type is the actual action object.

File: packages/data/src/registry.js [hunk 4]
```diff
@@ -217,7 +217,7 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	function registerStoreInstance( name, createStore ) {
 		if ( stores[ name ] ) {
 			// eslint-disable-next-line no-console
-			console.error( 'Store "' + name + '" is already registered.' );
+			console.info( 'Store "' + name + '" is already registered.' );
```
Comment (noahtallen) [inline packages/data/src/registry.js]: Drive-by change -- I was seeing this error in production, and I don't think we need to error (or even warn), as it's not a scenario the consumer needs to worry about. It's supported behavior after all! (And hard/impossible to prevent sometimes)

**Comment by noahtallen** [comment]
> Let's ensure that we don't inadvertently trigger a host of "didn't wait for Promise to resolve" errors where code dispatches and moves on.

I linked the package locally to another repo, and didn't see this error (and I don't see it in Gutenberg either), so I think we're good!

I verified the types as well:

Screenshot 2023-07-12 at 4 59 24 PM

The action itself:

Screenshot 2023-07-12 at 4 59 32 PM

File: packages/data/src/registry.js (hunk 4 above)
Comment (jsnajdr) [reply packages/data/src/registry.js]: Let's keep the `.error` call there because it's intentional and carefully chosen 🙂 There are many similar places that also register things and also report an `.error`. We want to be consistent:
- Block X is already registered.
- Plugin X is already registered.
- Format X is already registered.

**Comment by jsnajdr** [comment]
> is this just a way to make the API consistent with async thunks?

The `Promise.resolve` wrapper has no other purpose than to be backward compatible. Somewhere in ancient past, when there were no thunks and async actions were implemented with `rungen` generators, this convention that action dispatch always returns a `Promise` was added.

It's actually a bad idea to do this, because it's quite common that action dispatches are completely synchronous. And yet even these sync dispatches return a meaningless `Promise`. But if we removed the wrapper, code that does `registry.dispatch( store ).foo().then( ... )` would start crashing because there would be no `.then`. That's the only thing that prevents us from removing this.

By the way, `await` doesn't have this problem because it automatically promisifies (also with `Promise.resolve`) the awaited values.

File: packages/data/src/types.ts (hunk 2 above)
Comment (jsnajdr) [reply packages/data/src/types.ts]: The convention is that plain Redux action dispatch indeed returns the action object itself, but for thunks the return value is the value returned from the thunk function. And generally, any Redux middleware that enhances the `dispatch` function can change the return type arbitrarily. In practice all middlewares only _add_ new possible return types to the existing one because they want to be compatible with each other.

> Can we tell if the action is a thunk?

In principle it should be possible. For the action creators in the `actions` field of the store config, we know what they return. Either a plain object or a thunk function. And we can infer types from that.

`redux-thunk` has types like `ThunkDispatch` and `ThunkAction` and uses them to correctly type dispatches:

https://github.com/reduxjs/redux-thunk/blob/master/src/types.ts#L14-L37

File: packages/data/src/registry.js (hunk 4 above)
Comment (tyxla) [reply packages/data/src/registry.js]: I tend to agree with Jarda. This error is a pretty big deal and we want to make sure for the developer to see it.

**Review by tyxla — APPROVED** [review]
I agree with @jsnajdr that we should keep the error (and not change it to a warning/info) when a store is already registered, but other than that, this PR is good to go IMHO 🚀

File: packages/data/src/registry.js (hunk 4 above)
Comment (noahtallen) [reply packages/data/src/registry.js]: Alright, I will remove it from this PR, but I'd like to continue discussing this :)

Comment (jsnajdr) [reply packages/data/src/registry.js]: To me it doesn't particularly matter whether these logs are `.error`, `.warn` or `.info`, but they should be consistent among various registries we have.

**Comment by noahtallen** [comment]
I spent some time testing this in a separate repo which has more Typescript stores. I improved the return type to correctly handle thunks (and async thunks). When the action creator's return type is another function, we then get the return type of the inner function, and use `Awaited` to extract the async return type. So a thunk like `const thunk = () => async () => 123` has the return type of `Promise<number>`, and not `Promise< ()=>Promise<number> >` or `Promise<Promise<number>>`

I created a few simple dummy actions and thunks to test the types with various combinations of arguments and async scenarios:

```ts
// Action creators with and without actions:
const testAction = ( foo: string ) => ( { type: 'FOO', foo } );
const testAction2 = () => ( { type: 'BAR' } );

// Various thunks with combos of async and arguments:
const testThunk =
	() =>
	( { dispatch } ) => {
		return 1234;
	};

const testThunk1 = ( test: string ) => (): CheckoutPaymentMethodSlug => {
	return 1 > 2 ? 'alipay' : 'apple-pay';
};

const testThunk2 = ( test: number ) => async () => {
	return Promise.resolve( 'Hello world' );
};

const testThunk3 = ( foo: number, bar: string ) => async ( { dispatch } ) => {
	return foo + bar;
};

// Elsewhere:
const {
	testAction,
	testAction2,
	testThunk,
	testThunk1,
	testThunk2,
	testThunk3,
} = useDispatch( testStore );

// typeof value is { type: string, foo: string }
testAction( 'hi' ).then( ( value ) => {} );

// typeof value is { type: string }
testAction2().then( ( value ) => {} );

// typeof value is number
testThunk().then( ( value ) => {} );

// typeof value is CheckoutPaymentMethodSlug
testThunk1( 'hi' ).then( ( value ) => {} );

// typeof value is string (importantly, NOT Promise<string>)
testThunk2( 1 ).then( ( value ) => {} );

// typeof value is string
testThunk3( 1, 'hello' ).then( ( value ) => {} );
```

File: packages/data/src/dispatch.ts [hunk 5]
```diff
@@ -28,8 +23,10 @@ import defaultRegistry from './default-registry';
  * ```
  * @return Object containing the action creators.
  */
-export function dispatch< T extends StoreDescriptor< AnyConfig > >(
-	storeNameOrDescriptor: string | T
-): ActionCreatorsOf< ConfigOf< T > > {
```
Comment (noahtallen) [inline packages/data/src/dispatch.ts:33]: While testing in another repo, I came across an error where `dispatch( 'store-name' )` had the return type `ActionCreatorsOf...`. This return type doesn't work, because we aren't using the object format of the store name. `useDispatch` handles this correctly (the return type should be unknown or similar, since we don't know any metadata about the store), so this basically makes it more similar to those types.

File: packages/data/src/registry.js (hunk 4 above)
Comment (noahtallen) [reply packages/data/src/registry.js]: Gotcha. My reasoning is about this line:

> This error is a pretty big deal and we want to make sure for the developer to see it.

Is it a big deal? We added it in the first place to handle scenarios where a store is imported by two separate javascript applications in the same context. (E.g. two different plugins.) If I have two different plugins (different codebases for the sake of argument) and each of them do `import store as fooStore from 'foo';`, then the registration function will get called twice.

Is there is anything for the plugin authors to do to improve the situation? WordPress data fully supports this scenario and everything works fine -- each plugin author just wanted access to the same store. Thankfully dependency extraction handles this for WordPress stores, but that may not be the case for everything!

That's why I don't agree with calling it `error`, because IMO that should mean something serious is happening. But I can understand why we want to keep things consistent. For the other messages like "Plugin X is already registered", are the scenarios actually bad enough to need intervention?

Comment (tyxla) [reply packages/data/src/registry.js]: My rationale was that it could be a big deal if you are hoping that importing the store again will re-initialize the store from scratch. Since this behavior is no longer true since #49134, there might be 3rd party usage that relied on this behavior and might have broken since, and without the error, developers will never know that something is going on. 

I guess, if you insist on making it a warning or an info message, it can still do a similar job, so I wouldn't strongly object, as long as we do it consistently as Jarda suggested.

Comment (noahtallen) [reply packages/data/src/registry.js]: That's a fair point. I think at this point I don't feel strongly enough to change it unless it comes up again in the future :)

**Comment by noahtallen** [comment]
Would appreciate a second review if folks have time, since I made some changes to account for thunks.

**Review by tyxla — APPROVED** [review]
Looking great 👍 

Thanks for adding extra comments 🚀
